internals/overlord/logstate/gatherer.go

```diff
@@ -132,18 +138,36 @@ func fillDefaultOptions(options *logGathererOptions) *logGathererOptions {
 
 // PlanChanged is called by the LogManager when the plan is changed, if this
 // gatherer's target exists in the new plan.
-func (g *logGatherer) PlanChanged(pl *plan.Plan, buffers map[string]*servicelog.RingBuffer) {
+// The labels map should not be modified while this method is running.
+func (g *logGatherer) PlanChanged(pl *plan.Plan, serviceData map[string]*ServiceData, labels map[string]map[string]string) {
+	// New plan - the labels may have changed. We should tell the main loop to
+	// flush the client now so that we don't later get out of sync, and send old
+	// logs with the new labels.
+	select {
+	// Check just in case the gatherer has already been stopped, so we don't get
+	// deadlocked here.
+	case <-g.tomb.Dying():
+		return
+	case g.flushCh <- struct{}{}:
+	}
+
 	// Remove old pullers
 	for _, svcName := range g.pullers.Services() {
 		svc, svcExists := pl.Services[svcName]
 		if !svcExists {
 			g.pullers.Remove(svcName)
+			if dying := g.setClientLabels(svcName, nil); dying {
```

Comment:
Nit: I think this would be simpler/clearer if we reversed the sense of the boolean so it was "did we set the labels?". And then there'd be no need for the named var IMO. "If that didn't go through, return":

```suggestion
			if !g.setClientLabels(svcName, nil) {
```

What do you think?

Reply:
I agree. Boolean return value is like answering a question. SetClientLabels -> true sounds like, yes it worked.
But I think the better option would have been to return an error, but that doesn't make much sense here.

For the most I actually think we don't care that it is dying. We just don't want it to block, so we can just drop the if statement entirely.